agent/lib_aws_sdk_php.c

```diff
@@ -0,0 +1,177 @@
+ * zend_eval_string("Aws\\Sdk::VERSION;", &retval, "Get AWS Version")
+ * By the time we have detected the existence of the aws-sdk-php and with
+ * default composer profject settings, it callable even from
+ * nr_aws_sdk_php_enable which will automatically load the class if it isn't
+ * loaded yet and then evaluate the string. However, in the rare case that files
+ * are not loaded via autoloader and/or have non-default composer classload
+ * settings, if the class is not found, PHP 8.2+ will generate a fatal
+ * unrecoverable uncatchable error error whenever it cannot find a class. While
+ * calling this from nr_aws_sdk_php_enable would have been great and would allow
+ * the sdk version value to be set only once, to avoid the very unlikely but not
+ * impossible fatal error, this will be called from the
+ * "Aws\\ClientResolver::_apply_user_agent" wrapper which GUARANTEES that
+ * aws/sdk exists and is already loaded.
+ *
+ *
+ * Additionally given that aws-sdk-php is currently detected from the
+ * AwsClient.php file, this method will always be called when a client is
+ * created unlike Sdk::construct which doesn't show with PHP 8.2+.
+ *
+ * Using Aws/Sdk::__construct for version is currently nonviable as it is
+ * unreliable as a version determiner.
+ * Having separate functionality to extract from Aws/Sdk::__construct
+ * is both not required and is redundant and causes additional overhead and
+ * so only one function is needed to extract version.
+ *
+ * Aws\\ClientResolver::_apply_user_agent a reliable function as it is
+ * always called on client initialization since it is key to populating
+ * the request headers, and it loads Sdk by default.
+ *
+ * Concerns about future/past proofing to the checking prioritized the following
+ * implementation vs using the eval method.
+ */
+void nr_lib_aws_sdk_php_handle_version() {
+  zval* zval_version = NULL;
+  zend_class_entry* class_entry = NULL;
+  char* version = NULL;
+
+  class_entry = nr_php_find_class("aws\\sdk");
+  if (NULL != class_entry) {
+    zval_version = nr_php_get_class_constant(class_entry, "VERSION");
+
+    if (nr_php_is_zval_non_empty_string(zval_version)) {
+      version = Z_STRVAL_P(zval_version);
+    }
+  }
+  if (NRINI(vulnerability_management_package_detection_enabled)) {
+    /* Add php package to transaction */
+    nr_txn_add_php_package(NRPRG(txn), PHP_PACKAGE_NAME, version);
+  }
+  nr_fw_support_add_package_supportability_metric(NRPRG(txn), PHP_PACKAGE_NAME,
+                                                  version);
+  nr_php_zval_free(&zval_version);
+}
+
+void nr_lib_aws_sdk_php_add_supportability_service_metric(
+    const char* service_name) {
+  int MAX_LEN = 512;
+  char buf[MAX_LEN];
+
+  if (NULL == service_name || '\0' == service_name[0]) {
+    return;
+  }
+  if (NULL == NRPRG(txn)) {
+    return;
+  }
+  buf[0] = '\0';
+
+  snprintf(buf, MAX_LEN, "%s%s", PHP_AWS_SDK_SERVICE_NAME_METRIC_PREFIX,
+           service_name);
```

Comment:
For efficiency reasons I suggest using strcpy/strcat instead of formatting function. Something like this should do the trick and will explicitly tell what is the maximum length of the AWS metric the agent will accept and report vs now I have to manually subtract length of `PHP_AWS_SDK_SERVICE_NAME_METRIC_PREFIX` from 512:
```suggestion
#define AWS_SERVICE_NAME_LEN_MAX 64 // maximum name of aws service arbitrarily selected 64
  char *buf = nr_alloca(sizeof(PHP_AWS_SDK_SERVICE_NAME_METRIC_PREFIX) + SERVICE_NAME_LEN_MAX);
  char *cp = buf;
  cp += nr_strcpy(cp, PHP_AWS_SDK_SERVICE_NAME_METRIC_PREFIX);
  cp += nr_strlcpy(cp, service_name, SERVICE_NAME_LEN_MAX);
```

Reply:
This is is similar to how some other metrics are created and has the advantage of automatically truncating for us.

Using str_cpy really doesn't seem more efficient.
Calling [nr_strcpy](https://github.com/newrelic/newrelic-php-agent/blob/main/axiom/util_strings.c#L152) adds multiple checks and call to strlen.
Then [nr_strlcpy](https://github.com/newrelic/newrelic-php-agent/blob/main/axiom/util_strings.c#L167) again does more checks and strlen function call before doing a memcopy.

Plus the size of the prefix and name need to be tracked.

Seems like other metric name generators in agent use `nr_formatf` but that creates a string that needs to be freed or snprint like [here](https://github.com/newrelic/newrelic-php-agent/blob/main/agent/php_api.c#L54).
snprintf seems to do what we want in this case.

Reply:
🤦 I always forget that `nr_` versions of `strcpy` have extra overhead because they call `strlen`. However, raw `strncpy` is an order of magnitude faster than `sprintf` as demonstrated by this simple program:
```
$ cat main.c
#include <stdio.h>
#include <string.h>
#include "nr_time.h"

#define METRIC_NAME_PREFIX "Supportability/Aws/Service/"

int main(int argc, char* argv[]) {
	int i;
	nrtime_t start, stop, sprintf_duration, strcpy_duration;
	int MAX_LEN = 512;
	char buf[MAX_LEN];
	char *cp;
	char *service_name = "DynamoDb";

       	start = nr_get_time();
	for (i = 0; i < 1000 * 1000; i++) {
		buf[0] = '\0';
		snprintf(buf, MAX_LEN, "%s%s", METRIC_NAME_PREFIX, service_name);
	}
       	stop = nr_get_time();
	sprintf_duration = stop - start;
	printf("%d x sprintf used " NR_TIME_FMT "us\n", i, sprintf_duration);

       	start = nr_get_time();
	for (i = 0; i < 1000 * 1000; i++) {
		cp = buf;
		strcpy(cp, METRIC_NAME_PREFIX);
		cp += sizeof(METRIC_NAME_PREFIX) - 1;
		strncpy(cp, service_name, 64);
	}
       	stop = nr_get_time();
	strcpy_duration = stop - start;
	printf("%d x strcpy used " NR_TIME_FMT "us\n", i, strcpy_duration);
	printf("sprintf vs strcpy diff " NR_TIME_FMT "us\n", sprintf_duration-strcpy_duration);
}
$ gcc main.c && ./a.out
1000000 x sprintf used 48124us
1000000 x strcpy used 5195us
sprintf vs strcpy diff 42929us
```
We should really consider these types of optimizations whenever we have a chance. I strongly suggest optimizing this piece of code. Just because something has been done in the past doesn't mean it needs to be repeated.

Reply:
The reason why `sprintf` is so much slower is that it has to parse the format string and the extract the correct argument from var arg list and cast it to correct type and then copy the memory from one place to another applying correct format. `sprintf` is useful under some circumstances, but this is a very simple case, where the whole machinery of `sprintf` is not needed.

Reply:
Good points.  
Please see: [09e54ad](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/09e54adffaa12b2e422122439c6b678dbbccf230)

Reply:
This is better, but as you said earlier `nr_str*` wrappers around `str` functions use `strlen`. This is an inefficiency and can be avoided here. Please replace `nr_strcpy` with raw (not `nr_` wrapped) `strcpy` and compute new value for `cp` after `strcpy` call. The sample program demonstrates this.

Reply:
please see [4c7896f](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/4c7896f506a833099c4fe7b6f6dbd5ab3873dd8e)

Reply:
It's better but still can be improved. There's no need for this computation (especially call to `nr_strlen`) at runtime: `MAX_METRIC_NAME_LEN - nr_strlen(PHP_AWS_SDK_SERVICE_NAME_METRIC_PREFIX)` on this line:
```
  strlcpy(
      cp, service_name,
      MAX_METRIC_NAME_LEN - nr_strlen(PHP_AWS_SDK_SERVICE_NAME_METRIC_PREFIX));
```
This can be pre-computed at compile time:
```
#define MAX_METRIC_NAME_LEN 256
#define MAX_AWS_SERVICE_NAME_LEN (MAX_METRIC_NAME_LEN - sizeof(PHP_AWS_SDK_SERVICE_NAME_METRIC_PREFIX))
...
strlcpy(cp, service_name, MAX_AWS_SERVICE_NAME_LEN);
```

Reply:
nice suggestion!  please see [842d437](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/842d4377f3be22d2feca2dd1f32e429ff4bd6a40)